### PR TITLE
FIR: SetterParameterTypeChecker

### PR DESCRIPTION
**Pull request by yantimirov-timur**
add WRONG_SETTER_PARAMETER_TYPE check to Fir

### CODE REVIEW

**Comment by demiurg906** [comment]
Please rebase your bracnh to master (there were some changes in reporting diagnostics) and cleanup commit history: squash all commits to one and remove merge commit

FYI: best way to update your branch to `master` is rebasing:
```
git fetch upstream master // upstream is a name of remote of JetBrains/kotlin repo
git rebase upstream/master
```

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirErrors.kt [hunk 1]
```diff
@@ -96,10 +96,16 @@ object FirErrors {
     val VAR_ANNOTATION_PARAMETER by error0<FirSourceElement, KtParameter>(SourceElementPositioningStrategies.VAL_OR_VAR_NODE)
 
     // Exposed visibility group
-    val EXPOSED_TYPEALIAS_EXPANDED_TYPE by error3<FirSourceElement, KtNamedDeclaration, FirEffectiveVisibility, FirMemberDeclaration, FirEffectiveVisibility>(SourceElementPositioningStrategies.DECLARATION_NAME)
-    val EXPOSED_FUNCTION_RETURN_TYPE by error3<FirSourceElement, KtNamedDeclaration, FirEffectiveVisibility, FirMemberDeclaration, FirEffectiveVisibility>(SourceElementPositioningStrategies.DECLARATION_NAME)
+    val EXPOSED_TYPEALIAS_EXPANDED_TYPE by error3<FirSourceElement, KtNamedDeclaration, FirEffectiveVisibility, FirMemberDeclaration, FirEffectiveVisibility>(
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirErrors.kt:99]: If you want to refactor some files according to code style then please do it in separate commit. Otherwise it's hard to understand where are meaningful changes and where is a clenup

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirErrors.kt [hunk 2]
```diff
@@ -164,9 +172,13 @@ object FirErrors {
     val USELESS_VARARG_ON_PARAMETER by warning0<FirSourceElement, KtParameter>()
 
     // Properties & accessors
-    val ABSTRACT_PROPERTY_IN_NON_ABSTRACT_CLASS by error2<FirSourceElement, KtModifierListOwner, FirMemberDeclaration, FirMemberDeclaration>(SourceElementPositioningStrategies.MODALITY_MODIFIER)
+    val ABSTRACT_PROPERTY_IN_NON_ABSTRACT_CLASS by error2<FirSourceElement, KtModifierListOwner, FirMemberDeclaration, FirMemberDeclaration>(
+        SourceElementPositioningStrategies.MODALITY_MODIFIER
+    )
     val PRIVATE_PROPERTY_IN_INTERFACE by error0<FirSourceElement, KtProperty>(SourceElementPositioningStrategies.VISIBILITY_MODIFIER)
 
+    val WRONG_SETTER_PARAMETER_TYPE by error0<FirSourceElement, KtProperty>(SourceElementPositioningStrategies.DEFAULT)
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirErrors.kt:180]: In FE 1.0 this diagnostic has two parameters with expected and actual setter parameter type. Please add them here too

See `Errors.WRONG_SETTER_PARAMETER_TYPE`

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirWrongSetterParameterTypeChecker.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        val valueType = valueParameter.returnTypeRef.coneType
+        val propertyType = declaration.returnTypeRef.coneType
+
+        if (valueType != propertyType) {
+            reporter.report(valueSource, WRONG_SETTER_PARAMETER_TYPE)
+        }
+    }
+}
+
+
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirWrongSetterParameterTypeChecker.kt:30]: Redundant line

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticFactoryDsl.kt [hunk 4]
```diff
@@ -8,6 +8,7 @@ package org.jetbrains.kotlin.fir.analysis.diagnostics
 import com.intellij.psi.PsiElement
 import org.jetbrains.kotlin.diagnostics.*
 import org.jetbrains.kotlin.fir.FirSourceElement
+import org.jetbrains.kotlin.psi.KtParameter
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticFactoryDsl.kt:11]: Redundant import

File: compiler/fir/analysis-tests/testData/resolveWithStdlib/callableReferences/varProperties.kt [hunk 5]
```diff
@@ -2,7 +2,7 @@ import kotlin.reflect.KMutableProperty0
 import kotlin.reflect.KMutableProperty1
 
 class A {
-    var bar: Int = 1
+    var bar:Int = 1
```
Comment (demiurg906) [inline compiler/fir/analysis-tests/testData/resolveWithStdlib/callableReferences/varProperties.kt:5]: Missing space (and also redundant chage)

File: compiler/fir/analysis-tests/testData/resolveWithStdlib/callableReferences/ifWithCR.kt [hunk 6]
```diff
@@ -5,4 +5,4 @@ fun main(x: Int) {
     val x = if (x > 1) x::readOnlyWrapper else x::mutableWrapper
 
     x.get()
-}
+}
```
Comment (demiurg906) [inline compiler/fir/analysis-tests/testData/resolveWithStdlib/callableReferences/ifWithCR.kt:8]: What is this change?

File: compiler/fir/analysis-tests/testData/resolveWithStdlib/contracts/fromSource/good/returnsImplies/propertyAccessors.kt [hunk 7]
```diff
@@ -27,4 +27,4 @@ fun test_1(a: A?) {
 fun test_2(a: A?) {
     a.isNotNull = true
     a.foo()
-}
\ No newline at end of file
+}
```
Comment (demiurg906) [inline compiler/fir/analysis-tests/testData/resolveWithStdlib/contracts/fromSource/good/returnsImplies/propertyAccessors.kt:30]: Same

**Comment by demiurg906** [comment]
Also please run `FirOldFrontendDiagnosticsTestGenerated` and `FirDiagnosticTestSpecGenerated`. There will be tests which need to update according to your changes

**Review by demiurg906 — CHANGES_REQUESTED** [review]

**Comment by jsjeon** [comment]
FYI, after https://github.com/JetBrains/kotlin/commit/b92dce9be4a515bfed57997381b62d1315c1409e, diagnostic list is automatically generated. You need to update the generator as well as IDE part by running:
```
$ ./gradlew :compiler:fir:checkers:generateCheckersComponents
$ ./gradlew :idea:idea-frontend-fir:generateCode
```

File: .idea/misc.xml [hunk 8]
```diff
@@ -12,7 +12,6 @@
       <item index="2" class="java.lang.String" itemvalue="org.gradle.api.tasks.options.Option" />
     </list>
   </component>
-  <component name="ExternalStorageConfigurationManager" enabled="true" />
```
Comment (mglukhikh) [inline .idea/misc.xml:15]: This change should not be here

File: compiler/fir/analysis-tests/testData/resolveWithStdlib/callableReferences/ifWithCR.kt (hunk 6 above)
Comment (mglukhikh) [reply compiler/fir/analysis-tests/testData/resolveWithStdlib/callableReferences/ifWithCR.kt:8]: This change should not be here. Please revert it (e.g. "Revert selected changes" from Git plugin).

File: compiler/fir/analysis-tests/testData/resolveWithStdlib/contracts/fromSource/good/returnsImplies/propertyAccessors.kt (hunk 7 above)
Comment (mglukhikh) [reply compiler/fir/analysis-tests/testData/resolveWithStdlib/contracts/fromSource/good/returnsImplies/propertyAccessors.kt:30]: This change should not be here. Please revert it (e.g. "Revert selected changes" from Git plugin).

**Review by mglukhikh — CHANGES_REQUESTED** [review]
The main changes look Ok for me, but irrelevant changes (see my comments) should be removed from the PR.

**Comment by mglukhikh** [comment]
Also, please rebase you changes on current master branch. At this moment the PR cannot be merged into master without resolving conflicts (see warning "This branch has conflicts that must be resolved" in the PR).

**Comment by demiurg906** [comment]
@yantimirov-timur Also please check [common rules](https://chris.beams.io/posts/git-commit/) of commit messages naming and update your commit accordingly to it

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt [hunk 9]
```diff
@@ -565,6 +566,7 @@ class FirDefaultErrorMessages : DefaultErrorMessages.Extension {
             map.put(PRIVATE_SETTER_FOR_ABSTRACT_PROPERTY, "Private setters are not allowed for abstract properties")
             map.put(PRIVATE_SETTER_FOR_OPEN_PROPERTY, "Private setters are not allowed for open properties")
             map.put(VAL_WITH_SETTER, "A 'val'-property cannot have a setter")
+            map.put(WRONG_SETTER_PARAMETER_TYPE, "Setter parameter type must be equal to the type of the property", TO_STRING, TO_STRING)
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:569]: The error message needs to render the (expected) property type.

https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/diagnostics/rendering/DefaultErrorMessages.java#L478

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors.WRONG_SETTER_PARAMETER_TYPE
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.types.coneType
+
+object FirGetterSetterChecker : FirPropertyChecker() {
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt:15]: Maybe `FirPropertyAccessorChecker` sounds better.

But, I'm not sure whether we need yet another checker. How about adding a routine to `FirDeclarationCheckerUtils#checkPropertyAccessors`: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt#L172 which will be invoked per property/accessors?

File: compiler/testData/diagnostics/tests/EnumEntryAsType.fir.kt [hunk 11]
```diff
@@ -10,7 +10,7 @@ enum class Color {
 class MyColor(val x: <!UNRESOLVED_REFERENCE!>Color.RED<!>, y: <!UNRESOLVED_REFERENCE!>Color.RED<!>) : <!UNRESOLVED_REFERENCE!>Color.RED<!> {
 
     var z: <!UNRESOLVED_REFERENCE!>Color.RED<!> = Color.RED
-    set(arg: <!UNRESOLVED_REFERENCE!>Color.RED<!>) { z = arg }
+    set(arg: <!UNRESOLVED_REFERENCE, WRONG_SETTER_PARAMETER_TYPE!>Color.RED<!>) { z = arg }
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/EnumEntryAsType.fir.kt:13]: This shouldn't be type mismatch: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/EnumEntryAsType.kt#L13

File: compiler/testData/diagnostics/tests/varargs/varargInSetter.fir.kt [hunk 12]
```diff
@@ -1,5 +1,5 @@
 class My {
-    var x: String = ""
+    var x: <!WRONG_SETTER_PARAMETER_TYPE!>String<!> = ""
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/varargs/varargInSetter.fir.kt:2]: This report seems not right either: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/varargs/varargInSetter.kt#L2

File: compiler/testData/diagnostics/tests/variance/InvariantPosition.fir.kt [hunk 13]
```diff
@@ -38,10 +38,10 @@ interface Test<in I, out O, P> {
     var neOk22: Inv<out O>
     var neOk23: Inv<out I>
 
-    var neOk30: <!WRONG_NUMBER_OF_TYPE_ARGUMENTS!>Pair<I, ><!>
+    var neOk30: <!WRONG_NUMBER_OF_TYPE_ARGUMENTS, WRONG_SETTER_PARAMETER_TYPE!>Pair<I, ><!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/variance/InvariantPosition.fir.kt:41]: All changes in this file look not correct: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/variance/InvariantPosition.kt#L41

Btw, these properties don't have user-contributed setters at all, which means you need to bail out for default accessors.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt (hunk 10 above)
Comment (yantimirov-timur) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt:15]: I'll add another diagnostics in this checker, e.g. GETTER_VISIBILITY_DIFFERS_FROM_PROPERTY_VISIBILITY, SETTER_VISIBILITY_INCONSISTENT_WITH_PROPERTY_VISIBILITY, etc. Therefore, I think, i need another checker.

**Comment by mglukhikh** [comment]
Everything is almost good now. However, you haven't updates `FirDiagnosticsList.kt`, so your code is incompilable. After updating this file, please run Generators > Generate FIR Checker Components + FIR/IDE diagnostics. Also, please delete these lines
```
-Removed redundant changes
-Removed old tests
-Add vararg condition and edit message
```

from commit message (this is not important).

**Review by mglukhikh — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt (hunk 10 above)
Comment (mglukhikh) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt:17]: Renamed to FirPropertyAccessorChecker manually

**Review by mglukhikh — APPROVED** [review]
Very well, I will do some additional tests and merge this PR. Thank you.

**Comment by mglukhikh** [comment]
Merged manually 👍

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt (hunk 10 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirGetterSetterChecker.kt:17]: Instead, I'm moving `FirDeclarationCheckerUtils#checkPropertyAccessors` to `FirPropertyAccessorChecker`: https://github.com/JetBrains/kotlin/pull/4279
